src/rascal/representations/calculator_spherical_invariants.hh

```diff
@@ -440,122 +527,77 @@ namespace rascal {
-            }
 
-            // Sum the gradients wrt the neighbour atom position
-            // compute the \grad_j p^{i} coeffs
-            for (auto neigh : center.pairs()) {
-              auto && atom_j = neigh.get_atom_j();
-              auto atom_j_tag = atom_j.get_atom_tag();
-              // compute grad contribution only if the neighbour is _not_ an
-              // image of the center
-              if (atom_j_tag == atom_i_tag) {
-                continue;
-              }
-              auto & grad_neigh_coefficients{
-                  expansions_coefficients_gradient[neigh]};
-              auto & soap_neigh_gradient{soap_vector_gradients[neigh]};
-
-              auto neigh_type = neigh.get_atom_type();
-              if ((neigh_type != spair_type[0]) and
-                  (neigh_type != spair_type[1])) {
-                // Save the cost of iteration
-                continue;
+      if (this->compute_gradients) {
+        const int atom_i_tag{center.get_atom_tag()};
+
+        // Sum the gradients wrt the neighbour atom position
+        // compute the \grad_i p^{k} coeffs where k is either i or j
+        for (auto neigh : center.pairs_with_self_pair()) {
+          auto && atom_j = neigh.get_atom_j();
+          const int atom_j_tag{atom_j.get_atom_tag()};
+          const bool is_center_atom{manager->is_center_atom(neigh)};
+          // compute grad contribution only if the neighbour is _not_ an
+          // image of the center
+          if (atom_j_tag == atom_i_tag and not is_center_atom) {
+            continue;
+          }
+          // c^{k}
+          auto & coefficients_j{expansions_coefficients[atom_j]};
+          std::vector<Key_t> keys_coef_j{coefficients_j.get_keys()};
+          // \grad_i c^{k}
+          auto & grad_neigh_coefficients{
+              expansions_coefficients_gradient[neigh]};
+          std::vector<Key_t> keys_coef_grad_neigh{
+              grad_neigh_coefficients.get_keys()};
+          // \grad_i p^{k}
+          auto & soap_neigh_gradient{soap_vector_gradients[neigh]};
+
+          // \grad_i p^{kab} = \grad_i c^{k a} c^{k b} + c^{k a} \grad_i c^{k b}
+          // by definition \grad_i c^{k a} is non zero for one key 'a' so
+          // either a == b and we compute one term with a factor of 2 or only
+          // one of the two terms is non zero hence the swap of entry when
+          // spair_type[0] > spair_type[1] == true
+          for (const auto & coef_key_1 : keys_coef_grad_neigh) {
+            // \grad_i c^{k a}
+            const auto & grad_neigh_coefficients_1{
+                grad_neigh_coefficients[coef_key_1]};
+
+            for (const auto & coef_key_2 : keys_coef_j) {
+              // c^{k b}
+              const auto & expansion_coefficients_j_2{
+                  coefficients_j[coef_key_2]};
+              bool sorted{true}, equal{false};
+              // make sure spair_type has sorted entries
+              if (coef_key_1[0] > coef_key_2[0]) {
```

Comment:
there is a `(a,b)` index to the power spectrum gradients with `a` and `b` atomic types but the coefs in `(a,b)` and `(b,a)` are the same so we only compute and keep  `(a,b)` with `a < b`. 
There are 3 cases in this loop resulting in different computation load:
+  `a < b` the first block
+  `a > b` the second block
+ `a == b` both 1st and 2nd block

so the `if` here are saving quite a bit of computation.  I don't think there is a point to precomputing the branching.

Reply:
Ok - is there (from heuristics) an order of those conditions which leads to the most common as the first, etc. Of course, it depends on the structure and the different atom types.
But given the experience, is there a more efficient order?

In the behler-implementation we do precompute those combinations.
It plays into my concern that the calculator does indeed adapt the structure further - but that also changes (possibly) how the data is stored. I'll make a longer issue out of it for future reference later.

Reply:
you are right we could do this shuffling ahead of time but I am not convinced it matters so much since the computation inside the ifs are 'heavy' and I am sure that it is a lot of work to do properly.